### PR TITLE
Ensure that types referenced by PowerShellStandard are present

### PR DESCRIPTION
**Pull request by JamesWTruher**
# PR Summary

I'm adding a set of tests which validates that the types which are present in PowerShellStandard are present in PowerShell Core. This is to be sure that we don't break folks which have been using PowerShell Standard when we release a new version of Core.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Review by daxian-dbw — APPROVED** [review]

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# This is a simple type check to validate that types in PowerShellStandard are present in System.Management.Automation.dll
+# It does not check member presence and info, just the properties of the type
+Describe "Types referenced by PowerShell Standard should not be missing" {
+    BeforeAll {
+        $assets = [io.path]::Combine("$PSScriptRoot", "assets", "standardtypes.csv")
```
Comment (iSazonov) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]:
```suggestion
        $assets = [IO.Path]::Combine("$PSScriptRoot", "assets", "standardtypes.csv")
```

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        # These are not members of the type
+        $typeProperties = "IsCollectible", "IsSZArray", "IsByRefLike", "IsConstructedGenericType", "IsGenericType", "IsGenericTypeDefinition",
+            "IsGenericParameter", "IsTypeDefinition", "IsSecurityCritical", "IsSecuritySafeCritical", "IsSecurityTransparent", "IsInterface",
+            "IsNested", "IsArray", "IsByRef", "IsPointer", "IsGenericTypeParameter", "IsGenericMethodParameter", "IsVariableBoundArray",
+            "IsAbstract", "IsImport", "IsSealed", "IsSpecialName", "IsClass", "IsNestedAssembly", "IsNestedFamANDAssem", "IsNestedFamily",
+            "IsNestedFamORAssem", "IsNestedPrivate", "IsNestedPublic", "IsNotPublic", "IsPublic", "IsAutoLayout", "IsExplicitLayout",
+            "IsLayoutSequential", "IsAnsiClass", "IsAutoClass", "IsUnicodeClass", "IsCOMObject", "IsContextful", "IsEnum", "IsMarshalByRef",
+            "IsPrimitive", "IsValueType", "IsSignatureType", "IsSerializable", "IsVisible"
+
+        $tests = Import-Csv $assets | %{
```
Comment (iSazonov) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]:
```suggestion
        $tests = Import-Csv $assets | ForEach-Object { 
```

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            "IsNestedFamORAssem", "IsNestedPrivate", "IsNestedPublic", "IsNotPublic", "IsPublic", "IsAutoLayout", "IsExplicitLayout",
+            "IsLayoutSequential", "IsAnsiClass", "IsAutoClass", "IsUnicodeClass", "IsCOMObject", "IsContextful", "IsEnum", "IsMarshalByRef",
+            "IsPrimitive", "IsValueType", "IsSignatureType", "IsSerializable", "IsVisible"
+
+        $tests = Import-Csv $assets | %{ 
+            @{ FullName = $_.FullName; TypeMetaData = $_ }
+            }
+    }
+
+    It "Type '<FullName>' should be present with correct attributes" -testcase $tests {
```
Comment (iSazonov) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]:
```suggestion
    It "Type '<FullName>' should be present with correct attributes" -TestCases $tests {
```

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        param ( $FullName, $TypeMetaData )
+        $t = [psobject].assembly.GetType($FullName)
+        $t | Should -Not -BeNullOrEmpty
+        foreach ( $property in $typeProperties ) {
+            if ( $typeMetaData.$property -ne $t.$property ) {
+                throw "$property value not correct"
```
Comment (iSazonov) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]:
```suggestion
                throw "$property value is not correct"
```

**Review by iSazonov — APPROVED** [review]
Some minor comment.

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# This is a simple type check to validate that types in PowerShellStandard are present in System.Management.Automation.dll
+# It does not check member presence and info, just the properties of the type
+Describe "Types referenced by PowerShell Standard should not be missing" {
```
Comment (iSazonov) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]:
```suggestion
Describe "Types referenced by PowerShell Standard should not be missing"  -Tags "CI" {
```

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 (hunk 1 above)
Comment (rjmholt) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]:
```suggestion
        $assets = [System.IO.Path]::Combine("$PSScriptRoot", "assets", "standardtypes.csv")
```

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    BeforeAll {
+        $assets = [io.path]::Combine("$PSScriptRoot", "assets", "standardtypes.csv")
+        # The properties of a type which should match PowerShell Standard
+        # These are not members of the type
+        $typeProperties = "IsCollectible", "IsSZArray", "IsByRefLike", "IsConstructedGenericType", "IsGenericType", "IsGenericTypeDefinition",
```
Comment (rjmholt) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]: Would be nicer to put these one-per-line for diffing/editing

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            "IsPrimitive", "IsValueType", "IsSignatureType", "IsSerializable", "IsVisible"
+
+        $tests = Import-Csv $assets | %{ 
+            @{ FullName = $_.FullName; TypeMetaData = $_ }
+            }
+    }
+
+    It "Type '<FullName>' should be present with correct attributes" -testcase $tests {
+        param ( $FullName, $TypeMetaData )
+        $t = [psobject].assembly.GetType($FullName)
```
Comment (rjmholt) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]:
```suggestion
        $type = [psobject].Assembly.GetType($FullName)
```

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            "IsGenericParameter", "IsTypeDefinition", "IsSecurityCritical", "IsSecuritySafeCritical", "IsSecurityTransparent", "IsInterface",
+            "IsNested", "IsArray", "IsByRef", "IsPointer", "IsGenericTypeParameter", "IsGenericMethodParameter", "IsVariableBoundArray",
+            "IsAbstract", "IsImport", "IsSealed", "IsSpecialName", "IsClass", "IsNestedAssembly", "IsNestedFamANDAssem", "IsNestedFamily",
+            "IsNestedFamORAssem", "IsNestedPrivate", "IsNestedPublic", "IsNotPublic", "IsPublic", "IsAutoLayout", "IsExplicitLayout",
+            "IsLayoutSequential", "IsAnsiClass", "IsAutoClass", "IsUnicodeClass", "IsCOMObject", "IsContextful", "IsEnum", "IsMarshalByRef",
+            "IsPrimitive", "IsValueType", "IsSignatureType", "IsSerializable", "IsVisible"
+
+        $tests = Import-Csv $assets | %{ 
+            @{ FullName = $_.FullName; TypeMetaData = $_ }
+            }
```
Comment (rjmholt) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]:
```suggestion
        }
```

**Comment by daxian-dbw** [comment]
@JamesWTruher CI builds failed and there are some feedback. Please take another look when you have time.

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,34 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (PoshChan) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]: @JamesWTruher, your last commit had 709 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ExecutionContext' should be present with correct attributes
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.158.0/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 27
27:         $t | Should -Not -BeNullOrEmpty
```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.OutputProcessingStateEventArgs' should be present with correct attributes
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.158.0/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 27
27:         $t | Should -Not -BeNullOrEmpty
```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.PSRemotingJob' should be present with correct attributes
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.158.0/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 27
27:         $t | Should -Not -BeNullOrEmpty
```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ActionPreference' should be present with correct attributes
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.158.0/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 27
27:         $t | Should -Not -BeNullOrEmpty
```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ActionPreferenceStopException' should be present with correct attributes
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.158.0/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 27
27:         $t | Should -Not -BeNullOrEmpty
```

Comment (PoshChan) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]: @JamesWTruher, your last commit had 709 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ExecutionContext' should be present with correct attributes
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 27
27:         $t | Should -Not -BeNullOrEmpty
```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.OutputProcessingStateEventArgs' should be present with correct attributes
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 27
27:         $t | Should -Not -BeNullOrEmpty
```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.PSRemotingJob' should be present with correct attributes
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 27
27:         $t | Should -Not -BeNullOrEmpty
```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ActionPreference' should be present with correct attributes
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 27
27:         $t | Should -Not -BeNullOrEmpty
```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ActionPreferenceStopException' should be present with correct attributes
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 27
27:         $t | Should -Not -BeNullOrEmpty
```

Comment (PoshChan) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]: @JamesWTruher, your last commit had 709 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ExecutionContext' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.OutputProcessingStateEventArgs' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.PSRemotingJob' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ActionPreference' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ActionPreferenceStopException' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```

Comment (PoshChan) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]: @JamesWTruher, your last commit had 709 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ExecutionContext' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.OutputProcessingStateEventArgs' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.PSRemotingJob' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ActionPreference' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ActionPreferenceStopException' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```

Comment (PoshChan) [inline test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]: @JamesWTruher, your last commit had 709 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ExecutionContext' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, /Users/runner/runners/2.164.6/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.OutputProcessingStateEventArgs' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, /Users/runner/runners/2.164.6/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.PSRemotingJob' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, /Users/runner/runners/2.164.6/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ActionPreference' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, /Users/runner/runners/2.164.6/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```
Types referenced by PowerShell Standard should not be missing.Type 'System.Management.Automation.ActionPreferenceStopException' should be present with correct attributes
```powershell
IsCollectible value is not correct
at <ScriptBlock>, /Users/runner/runners/2.164.6/work/1/s/test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1: line 30
30:                 throw "$property value is not correct"

```

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 (hunk 6 above)
Comment (JamesWTruher) [reply test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]: done!

File: test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1 (hunk 1 above)
Comment (JamesWTruher) [reply test/powershell/engine/Basic/StandardLibraryTypes.Tests.ps1]: done!

**Comment by JamesWTruher** [comment]
@daxian-dbw @TravisEz13 runs clean now

**Comment by TravisEz13** [comment]
@JamesWTruher Should this be back ported?

**Pull request merged**